### PR TITLE
TF to Torch: Demo of reusable functions to setup input and output feature testing

### PR DESCRIPTION
**Pull request by jimthompson5802**
# Code Pull Requests

While working on `test_sequence_features.py` developed two reusable functions that could facilitate testing input and output features.  These functions perform the following:
* generate synthetic data
* use  pre-processing methods to collect feature specific metadata
* augment the feature's definition with the collected metadata
* generate a suitable data structure that can be used by either the input feature's `forward()` method or the output `forward()` method (i.e., simulates what comes out of the combiner).

By using these functions, the developer does not have to deal with the mechanics of creating synthetic data in a suitable form and manipulating the feature's definition.  

The new functions are
* `tests/integration_tests.utils.setup_input_feature_test()`
* `tests/integration_tests.utils.setup_output_feature_test()`

 Two existing tests were adapted to use these new functions for purposes of demonstration:
* `tests/ludwig/features/test_binary_feature.py`
* `tests/ludwig/features/test_sequence_features.py`

Here is output from a test run with the new functions
```
platform linux -- Python 3.8.12, pytest-6.2.5, py-1.10.0, pluggy-1.0.0 -- /usr/local/bin/python
cachedir: .pytest_cache
rootdir: /opt/project, configfile: pytest.ini
plugins: timeout-2.0.1, cov-3.0.0, xdist-2.4.0, pycharm-0.7.0, forked-1.3.0, anyio-3.3.4
collected 19 items

../../tests/ludwig/features/test_binary_feature.py::test_binary_input_feature[passthrough] PASSED                                    [  5%]
../../tests/ludwig/features/test_binary_feature.py::test_binary_input_feature[dense] PASSED                                          [ 10%]
../../tests/ludwig/features/test_binary_feature.py::test_binary_output_feature[0] PASSED                                             [ 15%]
../../tests/ludwig/features/test_binary_feature.py::test_binary_output_feature[1] PASSED                                             [ 21%]
../../tests/ludwig/features/test_sequence_features.py::test_sequence_input_feature[SequenceInputFeature-embed] PASSED                [ 26%]
../../tests/ludwig/features/test_sequence_features.py::test_sequence_input_feature[SequenceInputFeature-rnn] PASSED                  [ 31%]
../../tests/ludwig/features/test_sequence_features.py::test_sequence_input_feature[SequenceInputFeature-parallel_cnn] PASSED         [ 36%]
../../tests/ludwig/features/test_sequence_features.py::test_sequence_input_feature[SequenceInputFeature-cnnrnn] PASSED               [ 42%]
../../tests/ludwig/features/test_sequence_features.py::test_sequence_input_feature[SequenceInputFeature-stacked_parallel_cnn] PASSED [ 47%]
../../tests/ludwig/features/test_sequence_features.py::test_sequence_input_feature[SequenceInputFeature-stacked_cnn] PASSED          [ 52%]
../../tests/ludwig/features/test_sequence_features.py::test_sequence_input_feature[SequenceInputFeature-transformer] PASSED          [ 57%]
../../tests/ludwig/features/test_sequence_features.py::test_sequence_input_feature[TextInputFeature-embed] PASSED                    [ 63%]
../../tests/ludwig/features/test_sequence_features.py::test_sequence_input_feature[TextInputFeature-rnn] PASSED                      [ 68%]
../../tests/ludwig/features/test_sequence_features.py::test_sequence_input_feature[TextInputFeature-parallel_cnn] PASSED             [ 73%]
../../tests/ludwig/features/test_sequence_features.py::test_sequence_input_feature[TextInputFeature-cnnrnn] PASSED                   [ 78%]
../../tests/ludwig/features/test_sequence_features.py::test_sequence_input_feature[TextInputFeature-stacked_parallel_cnn] PASSED     [ 84%]
../../tests/ludwig/features/test_sequence_features.py::test_sequence_input_feature[TextInputFeature-stacked_cnn] PASSED              [ 89%]
../../tests/ludwig/features/test_sequence_features.py::test_sequence_input_feature[TextInputFeature-transformer] PASSED              [ 94%]
../../tests/ludwig/features/test_sequence_features.py::test_sequence_output_feature PASSED                                           [100%]

====================================================== 19 passed, 6 warnings in 1.04s ======================================================
```

### CODE REVIEW

File: tests/integration_tests/utils.py [hunk 1]
```diff
@@ -683,3 +686,106 @@ def train_with_backend(
     finally:
         # Remove results/intermediate data saved to disk
         shutil.rmtree(output_dir, ignore_errors=True)
+
+
+# creates synthetic data and update feature definitions
```
Comment (skanjila) [inline tests/integration_tests/utils.py:691]: can we add python style comments here: https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html

File: tests/integration_tests/utils.py [hunk 2]
```diff
@@ -683,3 +686,106 @@ def train_with_backend(
...
+
+    # preprocess raw data to create feature specific tensors and generate metadata
+    metadata = {}
+    input_data, _, _, metadata = DataFramePreprocessor.preprocess_for_training(
+        [feature_definition],
+        training_set=df,
+        training_set_metadata=metadata
+    )
+
+    # update feature definitions with defaults where required
```
Comment (skanjila) [inline tests/integration_tests/utils.py:716]: This is really useful, are the defaults specified in a config, if not does it make sense to move the defaults to config

File: tests/integration_tests/utils.py [hunk 3]
```diff
@@ -683,3 +686,106 @@ def train_with_backend(
     finally:
         # Remove results/intermediate data saved to disk
         shutil.rmtree(output_dir, ignore_errors=True)
+
```
Comment (skanjila) [inline tests/integration_tests/utils.py:689]: just a general question so I understand a bit more context, are we saying these setup functions should be generic enough to be reused across the different types of features , do we need to refactor the other tests to use this?

File: tests/integration_tests/utils.py [hunk 4]
```diff
@@ -683,3 +686,106 @@ def train_with_backend(
...
+# This setup assumes no dependencies
+def setup_output_feature_test(
+        batch_size: int = None,
+        hidden_size: int = None,
+        seq_size: Optional[int] = None,
+        feature_definition: Dict = None,
+        feature_class: Type[OutputFeature] = None
+) -> Tuple:
+    # generate synthetic raw data for testing
+    np.random.seed(1919)
```
[the remark this answers is not in the document]
Comment (jimthompson5802) [reply tests/integration_tests/utils.py:746]: I set the seed for reproducible synthetic data.  The underlying functions use numpy functions for data generation.

When I did not have this set, the sequence input feature test would in different test cases based on the length of the generated sequence.  When the seed is set the failures were repeatable.

File: tests/integration_tests/utils.py [hunk 5]
```diff
@@ -683,3 +686,106 @@ def train_with_backend(
...
+# creates synthetic data and update feature definitions
+# This covers pre-processing functionality and
+# converting raw data to input specific tensors.
+def setup_input_feature_test(
+        batch_size: int = None,
+        feature_definition: Dict = None,
+        feature_class: Type[InputFeature] = None
+) -> Tuple:
+    # generate synthetic raw data for testing
+    np.random.seed(1919)
```
[the remark this answers is not in the document]
Comment (jimthompson5802) [reply tests/integration_tests/utils.py:700]: See above response.

File: tests/integration_tests/utils.py [hunk 6]
```diff
@@ -683,3 +686,106 @@ def train_with_backend(
...
+    dataset = build_synthetic_dataset(
+        batch_size,
+        [feature_definition]
+    )
+    raw_data = '\n'.join([str(r[0]) for r in dataset])
+    df = pd.read_csv(io.StringIO(raw_data))
```
[the remark this answers is not in the document]
Comment (jimthompson5802) [reply tests/integration_tests/utils.py:706]: When you write, "do any validations" what do you have in mind?  the data generated is synthetic and I'm using the same functions as ludwig's data synthesizer capability.  I'm trusting those functions work correctly.

File: tests/integration_tests/utils.py (hunk 2 above)
Comment (jimthompson5802) [reply tests/integration_tests/utils.py:716]: IIRC, the defaults are set as class attributes in each feature.

File: tests/integration_tests/utils.py (hunk 3 above)
Comment (jimthompson5802) [reply tests/integration_tests/utils.py:689]: I'm hoping the two setup functions can be used by any input or output feature test.   Some refactoring may be needed for the output setup when we complete the output sequence feature migration.

To use these functions in other feature unit tests, some refactoring will need to be done.  I'm hoping it will be a net decrease on LOC for the particular unit test.  

I think you may have worked on the `test_binary.py` test.  So you can see the difference.  Though in that particular one, there was only one test case, basically for the binary input feature.  I had to tweak the code in that.  I added the second function to test the binary output feature.

File: tests/integration_tests/utils.py (hunk 1 above)
Comment (jimthompson5802) [reply tests/integration_tests/utils.py]: Updated per comment.

File: ludwig/features/binary_feature.py [hunk 7]
```diff
@@ -116,6 +116,8 @@ class BinaryInputFeature(BinaryFeatureMixin, InputFeature):
     def __init__(self, feature, encoder_obj=None):
         super().__init__(feature)
         self.overwrite_defaults(feature)
+        if feature['encoder'] == 'dense':
+            feature['input_size'] = 1
```
[the remark this answers is not in the document]
Comment (jimthompson5802) [reply ludwig/features/binary_feature.py]: Yeah...you're correct.  The only reason I put the test is that the `passthrough` encoder does not care.  However, for consistency, I should just set `input_size` regardless of the encoder.

Set `input_size` to 1 regardless of encoder.

**Comment by jimthompson5802** [comment]
Binary unit test after addressing comments:
```
../../tests/ludwig/features/test_binary_feature.py::test_binary_input_feature[passthrough] PASSED        [ 25%]
../../tests/ludwig/features/test_binary_feature.py::test_binary_input_feature[dense] PASSED              [ 50%]
../../tests/ludwig/features/test_binary_feature.py::test_binary_output_feature[0] PASSED                 [ 75%]
../../tests/ludwig/features/test_binary_feature.py::test_binary_output_feature[1] PASSED                 [100%]

======================================== 4 passed, 6 warnings in 0.86s =========================================
```

**Review by justinxzhao — COMMENTED** [review]
I like the overall idea/direction of better feature IO testing with less duplicated code.

My main comment is that (I think) we could simplify this even more by:
1) fixing `output_shape` feature implementations to include the shape logic that's currently in unit tests
2) abstracting all shape checking to assert on the feature_obj itself (like `test_sequence_feature.py` -- left a comment about this)
3) expanding the scope of the "shared test code" to include shape asserts.

Here's one way this could look:

Input feature test:
```
def assert_input_feature_encoder_output(feature_to_test, feature_class, batch_size)
    # setup synthetic tensor and feature definition
    input_tensor, feature_definition = setup_input_feature_test(
        batch_size=batch_size,
        feature_definition=feature_to_test,
        feature_class=feature_class)

    # instantiate input feature object
    feature_obj = feature_class(feature_definition)

    # pass tensor through the input feature
    encoder_output = feature_obj(input_tensor)

    # confirm correctness of the encoder output
    assert isinstance(encoder_output, dict)
    assert 'encoder_output' in encoder_output
    assert isinstance(encoder_output['encoder_output'], torch.Tensor)

    # confirm encoder shape
    assert encoder_output['encoder_output'].shape == \
           (batch_size, *feature_obj.output_shape)
```

Input feature unit tests:
```
@pytest.mark.parametrize('enc_encoder', ['passthrough', 'dense'])
def test_binary_input_feature(enc_encoder: str) -> None:
    feature_to_test = binary_feature(encoder=enc_encoder, fc_size=FC_SIZE)
    assert_input_feature_encoder_output(feature_to_test, BinaryInputFeature, BATCH_SIZE)


@pytest.mark.parametrize('enc_encoder', ['passthrough', 'dense'])
def test_numerical_input_feature(enc_encoder: str) -> None:
    feature_to_test = numerical_feature(encoder=enc_encoder, fc_size=FC_SIZE)
    assert_input_feature_encoder_output(feature_to_test, NumericalInputFeature, BATCH_SIZE)
```

Output feature test:
```
def assert_output_feature_output(feature_to_test, batch_size, hidden_size, feature_class):
    input_for_output_feature, feature_definition = setup_output_feature_test(
        batch_size=batch_size,
        hidden_size=hidden_size,
        feature_definition=feature_to_test,
        feature_class=feature_class
    )

    # instantiate binary output feature object
    feature_obj = feature_class(feature_definition)

    # pass tensor through to the output feature
    # specify no dependencies
    feature_output = feature_obj(input_for_output_feature)

    # confirm correctness of the the binary encoder output structure
    assert isinstance(feature_output, dict)

    # confirm expected components of feature output
    assert 'logits' in feature_output
    assert feature_output['logits'].shape == (batch_size,)

    assert encoder_output['encoder_output'].shape == \
        (batch_size, *feature_obj.output_shape)
```

Output feature unit tests:
```
@pytest.mark.parametrize('num_fc_layers', [0, 1])
def test_binary_output_feature(num_fc_layers: int) -> None:
    feature_to_test = binary_feature(
        num_fc_layers=num_fc_layers,
        fc_size=FC_SIZE,
        loss={'type': 'cross_entropy'}
    )
    assert_output_feature_output(
        feature_to_test, BATCH_SIZE, HIDDEN_SIZE, BinaryOutputFeature)


@pytest.mark.parametrize(
    'loss_type',
    ['mean_squared_error', 'mean_absolute_error']
)
@pytest.mark.parametrize('num_fc_layers', [0, 1])
def test_numerical_output_feature(
        num_fc_layers: int,
        loss_type: str
) -> None:
    feature_to_test = numerical_feature(
        num_fc_layers=num_fc_layers,
        fc_size=FC_SIZE,
        loss={'type': loss_type}
    )

    assert_output_feature_output(
        feature_to_test, BATCH_SIZE, HIDDEN_SIZE, NumericalOutputFeature)
```
